File: python/examples/streaming.py
"""Streaming: live text deltas + tool_call / tool_result / usage / done events.

Mirrors js/examples/streaming.ts. Reads OPENROUTER_API_KEY from the environment
(never hardcoded, never printed); skips the live call if it is unset.

Run from a venv where the package is installed:
    OPENROUTER_API_KEY=... python examples/streaming.py
"""
from __future__ import annotations

import asyncio
import json
import os
import sys

from toolnexus import create_client, create_toolkit, define_tool


async def main() -> None:
    tk = await create_toolkit()

    def add(a: float, b: float) -> str:
        """Add two numbers and return the sum."""
        return str(int(a) + int(b)) if float(a).is_integer() and float(b).is_integer() else str(a + b)

    tk.register(
        define_tool(
            add,
            name="add",
            description="Add two numbers.",
            input_schema={
                "type": "object",
                "properties": {"a": {"type": "number"}, "b": {"type": "number"}},
                "required": ["a", "b"],
            },
        )
    )

    key = os.environ.get("OPENROUTER_API_KEY")
    if not key:
        print("no OPENROUTER_API_KEY — skipping (imports OK)")
        await tk.close()
        return

    agent = create_client(
        base_url="https://openrouter.ai/api/v1",
        style="openai",
        model=os.environ.get("OPENROUTER_MODEL", "openai/gpt-4o-mini"),
        api_key=key,
    )

    text_deltas = 0
    streamed_text = ""
    tool_call_ev = 0
    tool_result_ev = 0
    done = None

    sys.stdout.write("stream: ")
    sys.stdout.flush()
    async for ev in agent.stream(
        "Use the add tool to compute 21 + 21, then say the result in one short sentence.",
        tk,
    ):
        t = ev["type"]
        if t == "text":
            text_deltas += 1
            streamed_text += ev["delta"]
            sys.stdout.write(ev["delta"])
            sys.stdout.flush()
        elif t == "tool_call":
            tool_call_ev += 1
            print(f"\n[tool_call] {ev['name']}({json.dumps(ev['args'])})")
        elif t == "tool_result":
            tool_result_ev += 1
            print(f"[tool_result] {ev['name']} -> {ev['output']}")
        elif t == "usage":
            print(f"[usage] {json.dumps(ev['usage'])}")
        elif t == "done":
            done = ev["result"]

    await tk.close()

    print(
        f"\n\nsummary: text_deltas={text_deltas} tool_call={tool_call_ev} "
        f"tool_result={tool_result_ev} | done.tool_call_count="
        f"{getattr(done, 'tool_call_count', None)} turns={getattr(done, 'turns', None)} "
        f"usage={json.dumps(getattr(done, 'usage', {}))}"
    )
    ok = (
        text_deltas > 1
        and tool_call_ev >= 1
        and tool_result_ev >= 1
        and done is not None
        and done.usage["total_tokens"] > 0
        and "42" in done.text
    )
    if not ok:
        print("FAILED streaming did not produce deltas + tool events + final usage/answer")
        raise SystemExit(1)
    print("OK streaming verified: incremental text deltas, tool_call/tool_result events, final done with usage")


if __name__ == "__main__":
    asyncio.run(main())

File: js/examples/streaming.ts
/**
 * Streaming: live text deltas + tool_call / tool_result / usage / done events.
 *   OPENROUTER_API_KEY=... node --experimental-strip-types examples/streaming.ts
 */
import { createToolkit, defineTool, createClient } from "../dist/index.js"

const tk = await createToolkit({})
tk.register(
  defineTool({
    name: "add", description: "Add two numbers.",
    inputSchema: { type: "object", properties: { a: { type: "number" }, b: { type: "number" } }, required: ["a", "b"] },
    run: ({ a, b }) => `${Number(a) + Number(b)}`,
  }),
)

const KEY = process.env.OPENROUTER_API_KEY
if (!KEY) { console.log("no OPENROUTER_API_KEY — skipping"); process.exit(0) }
const agent = createClient({
  baseUrl: "https://openrouter.ai/api/v1", style: "openai",
  model: process.env.OPENROUTER_MODEL ?? "openai/gpt-4o-mini", apiKey: KEY,
})

let textDeltas = 0, streamedText = "", toolCallEv = 0, toolResultEv = 0
let done: any
process.stdout.write("stream: ")
for await (const ev of agent.stream("Use the add tool to compute 21 + 21, then say the result in one short sentence.", { toolkit: tk })) {
  if (ev.type === "text") { textDeltas++; streamedText += ev.delta; process.stdout.write(ev.delta) }
  else if (ev.type === "tool_call") { toolCallEv++; console.log(`\n[tool_call] ${ev.name}(${JSON.stringify(ev.args)})`) }
  else if (ev.type === "tool_result") { toolResultEv++; console.log(`[tool_result] ${ev.name} -> ${ev.output}`) }
  else if (ev.type === "usage") console.log(`[usage] ${JSON.stringify(ev.usage)}`)
  else if (ev.type === "done") done = ev.result
}
await tk.close()

console.log(`\n\nsummary: textDeltas=${textDeltas} toolCall=${toolCallEv} toolResult=${toolResultEv} | done.toolCallCount=${done?.toolCallCount} turns=${done?.turns} usage=${JSON.stringify(done?.usage)}`)
const ok = textDeltas > 1 && toolCallEv >= 1 && toolResultEv >= 1 && done && done.usage.totalTokens > 0 && done.text.includes("42")
if (!ok) { console.error("❌ streaming did not produce deltas + tool events + final usage/answer"); process.exit(1) }
console.log("✅ streaming verified: incremental text deltas, tool_call/tool_result events, final done with usage")
